### DISCUSSION TITLE
Extend Webhook Body so we don't need to querry the invoice anymore

### DISCUSSION THREAD

**Discussion by mp21-org** [btcpayserver/btcpayserver#3069] [Technical Support]
Since the Webhook payload is cryptographically signed, it should be safe to include more invoice details, thus making the call to getInvoice via the API redundant when processing an incomming payment. (SettledInvoice Event)

I for myself would need: InvoicePrice and OrderID and RefundEmail in order to process payments on the spot.

See the Mattermost discussion with @ndeet: https://chat.btcpayserver.org/btcpayserver/pl/3m1jprkspt8pidbyxui4go3r6a

**Comment by ndeet** [comment]
@petzsch thanks for bringing it up. So the goal is to avoid querying the invoice like with the old api but be able to process the order only by relying on validated payload data from the webhook sent. 

You mention "InvoicePrice" would this be the 1:1 invoice invoice currency price or is this the paid amount? What do you do if there was an overpayment? You would not know how much was paid. Or do you need the sum of all payments in invoice currency?

**Reply by mp21-org** [reply]
Good Point. For top-up invoices the invoicePrice gets updated to whatever the client actually paid.

For regular invoices, an overpayment is not handled at the moment. Though it would be nice to have the actually paid amount in the invoice currency available. But I think thats not even possible when querrying the invoice via a [Greenfield call](https://docs.btcpayserver.org/API/Greenfield/v1/#operation/Invoices_GetInvoices) at the moment.

**Comment by woutersamaey** [comment]
Honestly I'm not sure that I like this idea ... unless we take some special considerations in mind:
- Webhooks can fire quickly and data may be obsolete in a few ms.
- Webhooks can be re-sent upon delivery issues or even manually. This is fine, but the data needs to be accurate and not a resend of the old message.
- If the receiving party employs a shitty developer, who does not validate the signature, you're in trouble. But I guess you could be in trouble with the shitty developer no matter how you look at it.

Having just the invoice ID in the body forces you to load fresh data and think in this manner. It is safer.

For the Magento 2 module I have taken this approach. Only remember the ID and fetch everything fresh.

I kinda like the idea of fresh data...

As long as performance is not an issue, I think 1 extra call is not so bad.
